File: src/java/voldemort/store/versioned/InconsistencyResolvingStore.java
/*
 * Copyright 2008-2009 LinkedIn, Inc
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package voldemort.store.versioned;

import java.util.List;
import java.util.Map;

import voldemort.VoldemortException;
import voldemort.store.DelegatingStore;
import voldemort.store.Store;
import voldemort.store.StoreCapabilityType;
import voldemort.store.CompositeVoldemortRequest;
import voldemort.versioning.InconsistencyResolver;
import voldemort.versioning.Versioned;

/**
 * A Store that uses a InconsistencyResolver to eliminate some duplicates.
 * 
 * 
 */
/*
 * Note that unlike get and getAll, getVersions is not overridden so the
 * versions are not passed through the inconsistency resolver.
 */
public class InconsistencyResolvingStore<K, V, T> extends DelegatingStore<K, V, T> {

    private final InconsistencyResolver<Versioned<V>> resolver;

    public InconsistencyResolvingStore(Store<K, V, T> innerStore,
                                       InconsistencyResolver<Versioned<V>> resolver) {
        super(innerStore);
        this.resolver = resolver;
    }

    @Override
    public List<Versioned<V>> get(K key, T transforms) throws VoldemortException {
        return resolver.resolveConflicts(super.get(key, transforms));
    }

    @Override
    public Map<K, List<Versioned<V>>> getAll(Iterable<K> keys, Map<K, T> transforms)
            throws VoldemortException {
        Map<K, List<Versioned<V>>> m = super.getAll(keys, transforms);
        for(Map.Entry<K, List<Versioned<V>>> entry: m.entrySet())
            m.put(entry.getKey(), resolver.resolveConflicts(entry.getValue()));
        return m;
    }

    @Override
    public Object getCapability(StoreCapabilityType capability) {
        if(capability == StoreCapabilityType.INCONSISTENCY_RESOLVER)
            return this.resolver;
        else
            return super.getCapability(capability);
    }

    @Override
    public List<Versioned<V>> get(CompositeVoldemortRequest<K, V> request) throws VoldemortException {
        if(request.resolveConflicts()) {
            return resolver.resolveConflicts(super.get(request));
        }
        return super.get(request);
    }

    @Override
    public Map<K, List<Versioned<V>>> getAll(CompositeVoldemortRequest<K, V> request)
            throws VoldemortException {
        Map<K, List<Versioned<V>>> m = super.getAll(request);
        if(request.resolveConflicts()) {
            for(Map.Entry<K, List<Versioned<V>>> entry: m.entrySet()) {
                m.put(entry.getKey(), resolver.resolveConflicts(entry.getValue()));
            }
        }

        return m;
    }

}

File: clients/cpp/src/InconsistencyResolvingStore.cpp
/* -*- C++ -*-; c-basic-offset: 4; indent-tabs-mode: nil */
/*
 * Implementation for InconsistencyResolvingStore class.
 * 
 * Copyright (c) 2009 Webroot Software, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

#include "InconsistencyResolvingStore.h"
#include "voldemort/VoldemortException.h"


namespace Voldemort {

using namespace boost;

static const bool REPAIR_READS = true;

InconsistencyResolvingStore::
InconsistencyResolvingStore(shared_ptr<Store>& store)
    : substore(store) {

}

void InconsistencyResolvingStore::
addResolver(shared_ptr<InconsistencyResolver>& resolver) {
    chain.push_back(resolver);
}

std::list<VersionedValue>* InconsistencyResolvingStore::get(const std::string& key) {
    std::list<VersionedValue>* resultList = substore->get(key);
    if (resultList != NULL && resultList->size() > 1) {
        try {
            std::list<shared_ptr<InconsistencyResolver> >::const_iterator it;
            for (it = chain.begin(); 
                 it != chain.end() && resultList->size() > 1; ++it) {
                (*it)->resolveConflicts(resultList);
            }
        } catch (...) {
            delete resultList;
            throw;
        }
    }
    return resultList;
}

void InconsistencyResolvingStore::put(const std::string& key, 
                                       const VersionedValue& value) {
    return substore->put(key, value);
}

bool InconsistencyResolvingStore:: deleteKey(const std::string& key, 
                                             const Version& version) {
    return substore->deleteKey(key, version);
}

const std::string* InconsistencyResolvingStore::getName() {
    return substore->getName();
}

void InconsistencyResolvingStore::close() {
    substore->close();
}


} /* namespace Voldemort */
